source/extensions/filters/listener/proxy_protocol/proxy_protocol.cc

```diff
@@ -380,6 +404,25 @@ bool Filter::parseTlvs(const std::vector<uint8_t>& tlvs) {
           (*cb_->dynamicMetadata().mutable_filter_metadata())[metadata_key]);
       metadata.mutable_fields()->insert({key_value_pair->key(), metadata_value});
       cb_->setDynamicMetadata(metadata_key, metadata);
+      ENVOY_LOG(debug, fmt::format("XXX proxy_protocol: Adding TLV to metadata: metadata_key: {}, "
+                                   "key_value_pair->key(): {}, metadata: {}",
+                                   metadata_key, key_value_pair->key(), metadata.DebugString()));
+
+      // Save TLV to the filter state.
```

Comment:
note, i don't think we will get here if we are not actively saving TLV data to dynamic metadata (see [readExtensions](https://github.com/visualzhou/envoy/blob/dcea43689937104a1cc4875f71a1794bf7f6927d/source/extensions/filters/listener/proxy_protocol/proxy_protocol.cc#L439))

Reply:
maybe we need to add a new field to the proxy protocol protobuf [here](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/listener/proxy_protocol/v3/proxy_protocol.proto#extensions-filters-listener-proxy-protocol-v3-proxyprotocol)? something like `pass_through_tlvs`?

Reply:
Yeah, we will only get here if we save TLV data to dynamic metadata. Do you think we should just use the metadata to store the TLVs?

IIUC, `pass_through_tlvs` will be a boolean to indicate whether we should store the TLVs into the filter data?

Reply:
are you suggesting that we should make this a two step process, i.e., store the TLVs that we want to pass through in dynamic metadata and then use the `pass_through_tlvs` to gate whether or not we pass the values in the metadata to the filter data for processing by the network filter?

it would be nice to allow some sort of filter for the TLVs that are passed on, but maybe this would be a further addition to the `pass_through_tlvs` option? i.e., we add some list of matcher rules for TLV types and values specific to the `pass_through_tlvs` option separate from `rules.[on_tlv_present|tlv_type]` that can be used to filter the TLVs that are passed through? alternatively, we could add `pass_through_tlv` at the same level as `on_tlv_present` and overload `tlv_type` to take a wildcard value, but maybe that would be too invasive/wonky?

Reply:
I see. I was referring to separating storing dynamic metadata and storing the filter state. If we construct the filter data after parsing the extensions as another comment suggested, it's not an issue any more.

Speaking of the config, let's make it concrete. The two options I can think of are:

Merging with the dynamic metadata config.
```
    - name: downstream_proxy_protocol
      typed_config:
        "@type": type.googleapis.com/envoy.extensions.filters.listener.proxy_protocol.v3.ProxyProtocol
        rules:
        - tlv_type: 5
          on_tlv_present:
            # adding metadata_namespace here prevented extraction of the TLVs using metadata matcher path in the RBAC filter
            key: uuid
            pass_through: true    <----- the new field
```

Or separate from the dynamic metadata config at the same level as `rules`.
```
    - name: downstream_proxy_protocol
      typed_config:
        "@type": type.googleapis.com/envoy.extensions.filters.listener.proxy_protocol.v3.ProxyProtocol
        rules:
        - tlv_type: 5
          on_tlv_present:
            # adding metadata_namespace here prevented extraction of the TLVs using metadata matcher path in the RBAC filter
            key: uuid
        pass_through_tlvs:  # the tlv types we want to pass through
        - 5   # UUID
        - 234 # PP2_TYPE_AWS (0xEA) for the ID of the private link endpoint
```

Reply:
i think i like the second option better for a couple of reasons:
- we may not be interested in the dynamic metadata, so there should be a way to pass the TLV on without storing it in dynamic metadata
- we can make the second option a bit more expressive by extending it with include/exclude lists (mutually exclusive)

if we extend option 2 to have include/exclude lists, we can pass all TLVs along without having to list every TLV we are interested in or just ignore ones we know we are not interested in. similarly, we can target specific TLVs to pass through if we are only interested in those particular TLVs upstream.

pass through only TLVs with type 5 or 234 if they are present
```
    - name: downstream_proxy_protocol
      typed_config:
        "@type": type.googleapis.com/envoy.extensions.filters.listener.proxy_protocol.v3.ProxyProtocol
        rules:
        - tlv_type: 5
          on_tlv_present:
            key: uuid
        pass_through_tlvs:
        # pass through TLVs with type 5 or 234
        - filter_type: include
        - tlv_types:
                - 5   # UUID
                - 234 # PP2_TYPE_AWS (0xEA) for the ID of the private link endpoint
```

pass through all TLVs
```
    - name: downstream_proxy_protocol
      typed_config:
        "@type": type.googleapis.com/envoy.extensions.filters.listener.proxy_protocol.v3.ProxyProtocol
        rules:
        - tlv_type: 5
          on_tlv_present:
            key: uuid
        pass_through_tlvs:
        # pass through TLVs with type 5 or 234
        - filter_type: exclude
        - tlv_types: []
```